src/mod_mam_utils.erl

```diff
@@ -207,16 +207,31 @@ encode_compact_uuid(Microseconds, NodeId)
 
 %% @doc Extract date and node id from a message id.
 -spec decode_compact_uuid(integer()) -> {integer(), byte()}.
-decode_compact_uuid(Id) ->
+decode_compact_uuid(Id) when is_integer(Id) ->
```

Comment:
For one of our client we patched this to use timestamps as uuids.
Or anything else. The point was to introduce mam_uid, that would encode from/to compact uids.

```diff
commit 2873535835e4c1f2385253516e389cdae9f2d4b4
Author: Michael Uvarov <arcusfelis@gmail.com>
Date:   Fri Nov 25 14:15:11 2016 +0100

    Separate message id logic into mam_uid

diff --git a/apps/ejabberd/src/mam_uid.erl b/apps/ejabberd/src/mam_uid.erl
new file mode 100644
index 000000000..968f9f157
--- /dev/null
+++ b/apps/ejabberd/src/mam_uid.erl
@@ -0,0 +1,11 @@
+%% @doc MAM message unique ID
+%%
+%% How to generate or represent message ids
+%% This module is global, so it works for both mod_mam and mod_mam_muc
+-module(mam_uid).
+
+-callback generate_message_id() -> integer().
+-callback encode_compact_uuid(integer(), integer()) -> integer().
+-callback decode_compact_uuid(integer()) -> {integer(),byte()}.
+-callback mess_id_to_external_binary(integer()) -> binary().
+-callback external_binary_to_mess_id(binary()) -> integer().

diff --git a/apps/ejabberd/src/mam_uid_nodetime.erl b/apps/ejabberd/src/mam_uid_nodetime.erl
new file mode 100644
index 000000000..f55458d7e
--- /dev/null
+++ b/apps/ejabberd/src/mam_uid_nodetime.erl
@@ -0,0 +1,54 @@
+%% @doc Message id based on node_id and unique timestamp
+%%
+%% Message ids unique unless nodes get restarted
+%% We use HEX encoding for external ids
+%% Ids are not transarent for the clients (they can't get timestamp from id)
+-module(mam_uid_nodetime).
+-behaviour(mam_uid).
+
+%% UID
+-export([generate_message_id/0,
+         encode_compact_uuid/2,
+         decode_compact_uuid/1,
+         mess_id_to_external_binary/1,
+         external_binary_to_mess_id/1]).
+
+%% -----------------------------------------------------------------------
+%% UID
+
+-spec generate_message_id() -> integer().
+generate_message_id() ->
+    {ok, NodeId} = ejabberd_node_id:node_id(),
+    %% Use monotone function here.
+    encode_compact_uuid(mod_mam_utils:now_to_microseconds(now()), NodeId).
+
+
+%% @doc Create a message ID (UID).
+%%
+%% It removes a leading 0 from 64-bit binary representation.
+%% It puts node id as a last byte.
+%% The maximum date, that can be encoded is `{{4253,5,31},{22,20,37}}'.
+-spec encode_compact_uuid(integer(), integer()) -> integer().
+encode_compact_uuid(Microseconds, NodeId)
+    when is_integer(Microseconds), is_integer(NodeId) ->
+    (Microseconds bsl 8) + NodeId.
+
+%% @doc Extract date and node id from a message id.
+-spec decode_compact_uuid(integer()) -> {integer(),byte()}.
+decode_compact_uuid(Id) ->
+    Microseconds = Id bsr 8,
+    NodeId = Id band 255,
+    {Microseconds, NodeId}.
+
+
+%% @doc Encode a message ID to pass it to the user.
+-spec mess_id_to_external_binary(integer()) -> binary().
+mess_id_to_external_binary(MessID) when is_integer(MessID) ->
+    list_to_binary(integer_to_list(MessID, 32)).
+
+
+%% @doc Decode a message ID received from the user.
+-spec external_binary_to_mess_id(binary()) -> integer().
+external_binary_to_mess_id(BExtMessID) when is_binary(BExtMessID) ->
+    list_to_integer(binary_to_list(BExtMessID), 32).
diff --git a/apps/ejabberd/src/mod_mam.erl b/apps/ejabberd/src/mod_mam.erl
index e889b86f9..bc857675e 100644
--- a/apps/ejabberd/src/mod_mam.erl
+++ b/apps/ejabberd/src/mod_mam.erl
@@ -59,6 +59,8 @@
 
 %% for feature (escalus) tests
 -export([set_params/1]).
+%% Called from mod_mam_utils
+-export([uid_module/0]).
 
 %% ----------------------------------------------------------------------
 %% Imports
@@ -990,11 +992,13 @@ params_helper(Params) ->
         "is_complete_message() -> ~p.~n"
         "default_result_limit() -> ~p.~n"
         "max_result_limit() -> ~p.~n"
+        "uid_module() -> ~p.~n"
         "params() -> ~p.~n",
         [proplists:get_bool(add_archived_element, Params),
          proplists:get_value(is_complete_message, Params, mod_mam_utils),
          proplists:get_value(default_result_limit, Params, 50),
          proplists:get_value(max_result_limit, Params, 50),
+         proplists:get_value(uid_module, Params, mam_uid_nodetime),
          Params
         ]))).
 
@@ -1009,3 +1013,9 @@ add_archived_element() ->
 is_complete_message(Module, Dir, Packet) ->
     M = mod_mam_params:is_complete_message(),
     M:is_complete_message(Module, Dir, Packet).
+
+uid_module() ->
+    %% Ensure that it's still works in tests
+    try mod_mam_params:uid_module()
+    catch _:_ -> mam_uid_nodetime
+    end.


diff --git a/apps/ejabberd/src/mod_mam_utils.erl b/apps/ejabberd/src/mod_mam_utils.erl
index d5263aab9..faa392995 100644
--- a/apps/ejabberd/src/mod_mam_utils.erl
+++ b/apps/ejabberd/src/mod_mam_utils.erl
@@ -163,10 +163,8 @@ microseconds_to_datetime(MicroSeconds) when is_integer(MicroSeconds) ->
 
 -spec generate_message_id() -> integer().
 generate_message_id() ->
-    {ok, NodeId} = ejabberd_node_id:node_id(),
-    %% Use monotone function here.
-    encode_compact_uuid(now_to_microseconds(now()), NodeId).
-
+    M = mod_mam:uid_module(),
+    M:generate_message_id().
 
 %% @doc Create a message ID (UID).
 %%
@@ -174,23 +172,22 @@ generate_message_id() ->
 %% It puts node id as a last byte.
 %% The maximum date, that can be encoded is `{{4253,5,31},{22,20,37}}'.
 -spec encode_compact_uuid(integer(), integer()) -> integer().
-encode_compact_uuid(Microseconds, NodeId)
-    when is_integer(Microseconds), is_integer(NodeId) ->
-    (Microseconds bsl 8) + NodeId.
-
+encode_compact_uuid(Microseconds, NodeId) ->
+    M = mod_mam:uid_module(),
+    M:encode_compact_uuid(Microseconds, NodeId).

 %% @doc Extract date and node id from a message id.
 -spec decode_compact_uuid(integer()) -> {integer(),byte()}.
 decode_compact_uuid(Id) ->
-    Microseconds = Id bsr 8,
-    NodeId = Id band 255,
-    {Microseconds, NodeId}.
+    M = mod_mam:uid_module(),
+    M:decode_compact_uuid(Id).
 
 
 %% @doc Encode a message ID to pass it to the user.
 -spec mess_id_to_external_binary(integer()) -> binary().
-mess_id_to_external_binary(MessID) when is_integer(MessID) ->
-    list_to_binary(integer_to_list(MessID, 32)).
+mess_id_to_external_binary(MessID) ->
+    M = mod_mam:uid_module(),
+    M:mess_id_to_external_binary(MessID).
 
 
 -spec maybe_external_binary_to_mess_id(binary()) -> undefined | integer().
@@ -202,8 +199,9 @@ maybe_external_binary_to_mess_id(BExtMessID) ->
 
 %% @doc Decode a message ID received from the user.
 -spec external_binary_to_mess_id(binary()) -> integer().
-external_binary_to_mess_id(BExtMessID) when is_binary(BExtMessID) ->
-    list_to_integer(binary_to_list(BExtMessID), 32).
+external_binary_to_mess_id(BExtMessID) ->
+    M = mod_mam:uid_module(),
+    M:external_binary_to_mess_id(BExtMessID).
 
 %% -----------------------------------------------------------------------
 %% XML
```

We also added file that allows to return timestamp in the same format as ejabberd does if mod_mam is configured with {uid_module, mam_uid_timestamp} option.

```erlang
%% @doc Message id based on timestamp                                                
-module(mam_uid_timestamp).                                                          
-behaviour(mam_uid).                                                                 
                                                                                     
%% UID                                                                               
-export([generate_message_id/0,                                                      
         encode_compact_uuid/2,                                                      
         decode_compact_uuid/1,                                                      
         mess_id_to_external_binary/1,                                               
         external_binary_to_mess_id/1]).                                             
                                                                                     
%% -----------------------------------------------------------------------           
%% UID                                                                               
                                                                                     
-spec generate_message_id() -> integer().                                            
generate_message_id() ->                                                             
    mod_mam_utils:now_to_microseconds(jlib:unique_timestamp()).                      
                                                                                     
                                                                                     
%% @doc Create a message ID (UID).                                                   
-spec encode_compact_uuid(integer(), integer()) -> integer().                        
encode_compact_uuid(Microseconds, NodeId)                                            
    when is_integer(Microseconds), is_integer(NodeId) ->                             
    Microseconds.                                                                    
                                                                                     
                                                                                     
%% @doc Extract date and node id from a message id.                                  
-spec decode_compact_uuid(integer()) -> {integer(),byte()}.                          
decode_compact_uuid(Id) ->                                                           
    Microseconds = Id,                                                               
    NodeId = 0,                                                                      
    {Microseconds, NodeId}. 

%% @doc Encode a message ID to pass it to the user.                                  
-spec mess_id_to_external_binary(integer()) -> binary().                             
mess_id_to_external_binary(MessID) when is_integer(MessID) ->                        
    list_to_binary(integer_to_list(MessID)).                                         
                                                                                     
                                                                                     
%% @doc Decode a message ID received from the user.                                  
-spec external_binary_to_mess_id(binary()) -> integer().                             
external_binary_to_mess_id(BExtMessID) when is_binary(BExtMessID) ->                 
    list_to_integer(binary_to_list(BExtMessID)). 
```

I know, that this patch is NOT compatible with the latest MongooseIM.
But I would love to see uid_module introduced and guid-uuid stuff introduced as a configurable option.
Because there is always more than one way to encode these uids.

Reply:
Sorry if stupid question but why after all does the upper level need to know anything about UUIDs and convert them back and forth between external and internal representation?

Also in our scenario (we do not persist here at the moment) generate_message_id wouldn't work for us because IDs are created at some other external service (REST) when persisting a message and not before.

Apart from that I like the idea except for making configuration more complicated again. But maybe I misunderstood your approach. If not couldn't we go for some conventional naming here instead (eg: `mod_mam_cassandra_arch` -> `mod_mam_cassandra_uid`) or just make the interface part of the backend behaviour?

Reply:
We can pass stanza into generate_message_id as an argument.
Than we can extract UUID from the stanza itself.

We can return nothing from the function, if backend does not need it.

I just want to keep the current behaviour as the basic point and if some project needs to change something, it can create another module of mam_uid and keep the logic here.

To answer your question, we need to know UUID to allow async message writers to work.

About naming. I am just using the same mechanism as `mam_jid.erl`.

There is a file `mam_jid` that controls, how jids should be encoded in the database.
There are two modules:
- mam_jid_mini - something that was originally
- mam_jid_rfc - something that we decided to switch to.

The same with `mam_message` module:
- mam_message_xml
- mam_message_eterm
- mam_message_compressed_eterm

Our conversation is that only modules that should have `mod_` prefix should be modules that implement `gen_mod`. At least the new ones :)

Reply:
Maybe we can introduce `mam_uid` in second iteration of this improvement, as a part of our next sprint? And stick to Stefan's implementation for iteration 1?

Reply:
+1 for how you want to handle 'mod_' prefixes :p 

As for me I'm fine with merging of course but I can also offer to contribute to any more sophisticated solution. Be it within the scope of this PR or any other.

That said, I'm not working full time currently and will be 'off' for two weeks now. This does not include responses here but potential work on this or any other PR.

Reply:
We usually don't have time for iterations 2 :)
I just don't want to have _more_ logic behind ids by _default_.